File: packages/client-python/src/rocketride/cli/commands/store.py
# MIT License
#
# Copyright (c) 2026 Aparavi Software AG
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
File-store commands: ``store dir/type/write/rm/mkdir/stat``.

Thin wrappers over the client's ``fs_*`` methods with DOS-dir-style
listing output. Kept in exact parity with the TypeScript CLI's
``src/cli/commands/store.ts``.
"""

import sys
from datetime import datetime, timezone

from ..utils.common import connect_client, run_cli_command
from ..utils.output import Output


async def run_store(args) -> int:
    """
    Execute one ``store`` subcommand.

    Args:
        args: Parsed argparse namespace (store_subcommand, path, ...).

    Returns:
        Exit code.
    """

    async def action(out: Output) -> int:
        client = await connect_client(args.uri, args.apikey)
        subcommand = args.store_subcommand

        if subcommand == 'dir':
            # step: list a directory in DOS dir style
            path = getattr(args, 'path', '') or ''
            result = await client.fs_list_dir(path)
            entries = result.get('entries', [])
            if not entries:
                stat = await client.fs_stat(path) if path else {'exists': True, 'type': 'dir'}
                if stat.get('exists') and stat.get('type') == 'dir':
                    out.line(f'    {0:>8,} File(s)  {0:>14,} bytes')
                    out.line(f'    {0:>8,} Dir(s)')
                else:
                    out.line('File Not Found')
                out.result({'path': path, 'entries': []})
                return 0
            total_size = 0
            file_count = 0
            dir_count = 0
            for entry in entries:
                modified = entry.get('modified')
                if modified:
                    dt = datetime.fromtimestamp(modified, tz=timezone.utc)
                    date_str = dt.strftime('%m/%d/%Y  %I:%M %p')
                else:
                    date_str = '                   '
                if entry.get('type') == 'dir':
                    out.line(f'{date_str}    <DIR>          {entry["name"]}')
                    dir_count += 1
                else:
                    size = entry.get('size', 0) or 0
                    total_size += size
                    out.line(f'{date_str}    {size:>14,} {entry["name"]}')
                    file_count += 1
            out.line(f'    {file_count:>8,} File(s)  {total_size:>14,} bytes')
            out.line(f'    {dir_count:>8,} Dir(s)')
            out.result({'path': path, 'entries': entries})
            return 0

        if subcommand == 'type':
            # step: stream the file's text to stdout (raw, no decoration)
            text = await client.fs_read_string(args.path)
            if not out.json_requested:
                sys.stdout.write(text)
            out.result({'path': args.path, 'content': text})
            return 0

        if subcommand == 'write':
            # step: upload a local file through the handle API, or write inline text
            if getattr(args, 'file', None):
                info = await client.fs_open(args.path, 'w')
                handle = info['handle']
                try:
                    with open(args.file, 'rb') as f:
                        while True:
                            chunk = f.read(65536)
                            if not chunk:
                                break
                            await client.fs_write(handle, chunk)
                    await client.fs_close(handle, 'w')
                except Exception:
                    try:
                        await client.fs_close(handle, 'w')
                    except Exception:  # noqa: BLE001
                        pass
                    raise
            elif getattr(args, 'content', None) is not None:
                await client.fs_write_string(args.path, args.content)
            else:
                return out.fail('Either --file or --content is required')
            out.line(f'Written: {args.path}')
            out.result({'path': args.path, 'written': True})
            return 0

        if subcommand == 'rm':
            await client.fs_delete(args.path)
            out.line(f'Deleted: {args.path}')
            out.result({'path': args.path, 'deleted': True})
            return 0

        if subcommand == 'mkdir':
            await client.fs_mkdir(args.path)
            out.line(f'Created: {args.path}/')
            out.result({'path': args.path, 'created': True})
            return 0

        if subcommand == 'stat':
            result = await client.fs_stat(args.path)
            if not result.get('exists'):
                out.line(f'{args.path}: not found')
            else:
                details = []
                if result.get('size') is not None:
                    details.append(f'size: {result["size"]:,}')
                if result.get('modified'):
                    modified = datetime.fromtimestamp(result['modified'], tz=timezone.utc)
                    details.append(f'modified: {modified.isoformat()}')
                suffix = f' ({", ".join(details)})' if details else ''
                out.line(f'{args.path}: {result.get("type")}{suffix}')
            out.result({'path': args.path, **result})
            return 0

        return out.fail(f'Unknown store subcommand: {subcommand}')

    return await run_cli_command(args, action)

File: packages/client-typescript/src/cli/commands/store.ts
/**
 * MIT License
 *
 * Copyright (c) 2026 Aparavi Software AG
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
 * File-store commands: `store dir/type/write/rm/mkdir/stat`.
 *
 * Thin wrappers over the client's fs* methods with DOS-dir-style listing
 * output (kept byte-compatible with the Python CLI's `store dir`).
 */

import * as fs from 'fs';
import { Command } from 'commander';
import { addConnectionOptions, connectClient, runCliCommand } from '../common';

/**
 * Register the `store` command group on the program.
 *
 * @param program - The root commander program.
 */
export function registerStoreCommands(program: Command): void {
	const storeCmd = program.command('store').description('File store operations');

	// ── store dir ────────────────────────────────────────────────────────
	const dirCmd = storeCmd
		.command('dir [path]')
		.description('List directory contents')
		.action(async (dirPath, options) => {
			await runCliCommand(options, async (out) => {
				const client = await connectClient(options);
				const result = await client.fsListDir(dirPath || '');
				const entries = result.entries || [];
				if (entries.length === 0) {
					const stat = dirPath ? await client.fsStat(dirPath) : { exists: true, type: 'dir' as const };
					if (stat.exists && stat.type === 'dir') {
						out.line(`    ${(0).toLocaleString().padStart(8)} File(s)  ${(0).toLocaleString().padStart(14)} bytes`);
						out.line(`    ${(0).toLocaleString().padStart(8)} Dir(s)`);
					} else {
						out.line('File Not Found');
					}
					out.result({ path: dirPath || '', entries: [] });
					return 0;
				}
				let totalSize = 0;
				let fileCount = 0;
				let dirCount = 0;
				for (const e of entries) {
					let dateStr = '                   ';
					if (e.modified) {
						const d = new Date(e.modified * 1000);
						const mm = String(d.getUTCMonth() + 1).padStart(2, '0');
						const dd = String(d.getUTCDate()).padStart(2, '0');
						const yyyy = d.getUTCFullYear();
						let hh = d.getUTCHours();
						const min = String(d.getUTCMinutes()).padStart(2, '0');
						const ampm = hh >= 12 ? 'PM' : 'AM';
						hh = hh % 12 || 12;
						dateStr = `${mm}/${dd}/${yyyy}  ${String(hh).padStart(2, '0')}:${min} ${ampm}`;
					}
					if (e.type === 'dir') {
						out.line(`${dateStr}    <DIR>          ${e.name}`);
						dirCount++;
					} else {
						const size = e.size ?? 0;
						totalSize += size;
						out.line(`${dateStr}    ${size.toLocaleString().padStart(14)} ${e.name}`);
						fileCount++;
					}
				}
				out.line(`    ${fileCount.toLocaleString().padStart(8)} File(s)  ${totalSize.toLocaleString().padStart(14)} bytes`);
				out.line(`    ${dirCount.toLocaleString().padStart(8)} Dir(s)`);
				out.result({ path: dirPath || '', entries });
				return 0;
			});
		});
	addConnectionOptions(dirCmd);

	// ── store type ───────────────────────────────────────────────────────
	const typeCmd = storeCmd
		.command('type <path>')
		.description('Display file contents')
		.action(async (filePath, options) => {
			await runCliCommand(options, async (out) => {
				const client = await connectClient(options);
				const text = await client.fsReadString(filePath);
				if (!out.jsonRequested) {
					process.stdout.write(text);
				}
				out.result({ path: filePath, content: text });
				return 0;
			});
		});
	addConnectionOptions(typeCmd);

	// ── store write ──────────────────────────────────────────────────────
	const writeCmd = storeCmd
		.command('write <path>')
		.description('Write a file')
		.option('--file <localFile>', 'Local file to upload')
		.option('--content <text>', 'Inline text content')
		.action(async (filePath, options) => {
			await runCliCommand(options, async (out) => {
				if (!options.file && options.content === undefined) {
					return out.fail('Either --file or --content is required');
				}
				const client = await connectClient(options);
				if (options.file) {
					// Stream the local file through the store handle API
					const { handle } = await client.fsOpen(filePath, 'w');
					try {
						const stream = fs.createReadStream(options.file);
						for await (const chunk of stream) {
							await client.fsWrite(handle, chunk as Uint8Array);
						}
						await client.fsClose(handle, 'w');
					} catch (err) {
						try {
							await client.fsClose(handle, 'w');
						} catch {
							/* best-effort */
						}
						throw err;
					}
				} else {
					await client.fsWriteString(filePath, options.content);
				}
				out.line(`Written: ${filePath}`);
				out.result({ path: filePath, written: true });
				return 0;
			});
		});
	addConnectionOptions(writeCmd);

	// ── store rm ─────────────────────────────────────────────────────────
	const rmCmd = storeCmd
		.command('rm <path>')
		.description('Delete a file')
		.action(async (filePath, options) => {
			await runCliCommand(options, async (out) => {
				const client = await connectClient(options);
				await client.fsDelete(filePath);
				out.line(`Deleted: ${filePath}`);
				out.result({ path: filePath, deleted: true });
				return 0;
			});
		});
	addConnectionOptions(rmCmd);

	// ── store mkdir ──────────────────────────────────────────────────────
	const mkdirCmd = storeCmd
		.command('mkdir <path>')
		.description('Create a directory')
		.action(async (dirPath, options) => {
			await runCliCommand(options, async (out) => {
				const client = await connectClient(options);
				await client.fsMkdir(dirPath);
				out.line(`Created: ${dirPath}/`);
				out.result({ path: dirPath, created: true });
				return 0;
			});
		});
	addConnectionOptions(mkdirCmd);

	// ── store stat ───────────────────────────────────────────────────────
	const statCmd = storeCmd
		.command('stat <path>')
		.description('Get file/directory metadata')
		.action(async (filePath, options) => {
			await runCliCommand(options, async (out) => {
				const client = await connectClient(options);
				const result = await client.fsStat(filePath);
				if (!result.exists) {
					out.line(`${filePath}: not found`);
				} else {
					const details: string[] = [];
					if (result.size !== undefined) details.push(`size: ${result.size.toLocaleString()}`);
					if (result.modified) details.push(`modified: ${new Date(result.modified * 1000).toISOString()}`);
					out.line(`${filePath}: ${result.type}${details.length ? ` (${details.join(', ')})` : ''}`);
				}
				out.result({ path: filePath, ...result });
				return 0;
			});
		});
	addConnectionOptions(statCmd);
}
